src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs

```diff
@@ -89,6 +96,18 @@ internal async Task<T> ExecuteRequestAsync<T>(Uri endPoint, HttpMethod method, R
 
                 if (response.StatusCode != HttpStatusCode.OK || expectErrorsOn200OK)
                 {
+                    //Check if we can perform device auth
+                    if (sendDeviceAuthParams)
+                    {
+                        string responseHeader = await _deviceAuthManager.CreateDeviceAuthChallengeResponseAsync(response, endpointUri).ConfigureAwait(false);
```

Comment:
I read throught the implementation and it seems to be ok for .net classic, but not ok for .net core, android etc. as the logic in `DeviceAuthHelper.GetBypassChallengeResponse` (in NullDeviceAuthManager` does not make any verification if the headers are there. 

In the .net 45 implementation you use `IsDeviceAuthChallenge` which is pretty defensive, but in the NullDeviceAuthManger you no longer use this.

Reply:
Also I would like to suggest a slightly different pattern to make the code more readable. The method `CreateXXX` implies that a response header will always be created. But in reality, this is not the case. 

So instead you can have a method that is more like `bool TryCreateXXX(..., out responseHeader)`. This is typical pattern in C#, see for example [here](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.trygetvalue?view=netframework-4.8).